## ABOUT THE PROYECT

## DATA EXTRACTION

In [68]:
import pandas as pd
import numpy as np

In [69]:
df2017=pd.read_csv('data/Produccion_Pesquera_2017.csv',encoding='latin-1')
df2018=pd.read_csv('data/Produccion_Pesquera_2018.csv',encoding='latin-1')
df2019=pd.read_csv('data/Produccion_Pesquera_2019.csv',encoding='latin-1')
df2020=pd.read_csv('data/Produccion_Pesquera_2020.csv',encoding='latin-1')
df=pd.concat([df2017, df2018,df2019,df2020],axis=0)
display(df.head(20))


C:\Users\fer-h\AppData\Local\Temp\ipykernel_14480\2379586494.py:1: DtypeWarning: Columns (1,2,3,4,5,6,8,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df2017=pd.read_csv('data/Produccion_Pesquera_2017.csv',encoding='latin-1')
C:\Users\fer-h\AppData\Local\Temp\ipykernel_14480\2379586494.py:2: DtypeWarning: Columns (1,2,3,4,5,6,8,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df2018=pd.read_csv('data/Produccion_Pesquera_2018.csv',encoding='latin-1')
C:\Users\fer-h\AppData\Local\Temp\ipykernel_14480\2379586494.py:3: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df2019=pd.read_csv('data/Produccion_Pesquera_2019.csv',encoding='latin-1')


EJERCICIO FISCAL   ENTIDAD FEDERATIVA         NOMBRE OFICINA MES DE CORTE  \
0             2017.0      BAJA CALIFORNIA             SAN FELIPE        JULIO   
1             2017.0      BAJA CALIFORNIA             SAN FELIPE        JULIO   
2             2017.0  BAJA CALIFORNIA SUR         BAHIA ASUNCION       AGOSTO   
3             2017.0  BAJA CALIFORNIA SUR          SANTA ROSALIA    NOVIEMBRE   
4             2017.0               COLIMA                TECOMAN      FEBRERO   
5             2017.0             CAMPECHE                 ATASTA    NOVIEMBRE   
6             2017.0             CAMPECHE               CAMPECHE        ENERO   
7             2017.0             CAMPECHE               CAMPECHE        ABRIL   
8             2017.0             CAMPECHE               CAMPECHE         MAYO   
9             2017.0             CAMPECHE              CHAMPOTON         MAYO   
10            2017.0             CAMPECHE              CHAMPOTON        JUNIO   
11            2017.0             CAMPECHE             ISLA ARENA   SEPTIEMBRE   
12            2017.0             CAMPECHE             ISLA ARENA    NOVIEMBRE   
13            2017.0             CAMPECHE               SABANCUY      OCTUBRE   
14            2017.0             CAMPECHE             SEYBAPLAYA        ENERO   
15            2017.0             GUERRERO  MARQUELIA CRUZ GRANDE    NOVIEMBRE   
16            2017.0             GUERRERO  MARQUELIA CRUZ GRANDE    NOVIEMBRE   
17            2017.0              JALISCO        PUERTO VALLARTA      FEBRERO   
18            2017.0              JALISCO        PUERTO VALLARTA        ABRIL   
19            2017.0              JALISCO        PUERTO VALLARTA         MAYO   

     ORIGEN NOMBRE COMUN PESO DESEMBARCADO KILOGRAMOS PESO VIVO KILOGRAMOS  \
0   CAPTURA      ABADEJO                           94                 94.0   
1   CAPTURA      ABADEJO                         1400               1540.0   
2   CAPTURA      ABADEJO                         4397               4397.0   
3   CAPTURA      ABADEJO                           17                 17.0   
4   CAPTURA      ABADEJO                         4587               4587.0   
5   CAPTURA      ABADEJO                         4406               4406.0   
6   CAPTURA      ABADEJO                           18                 18.0   
7   CAPTURA      ABADEJO                          103                103.0   
8   CAPTURA      ABADEJO                          189                189.0   
9   CAPTURA      ABADEJO                           52                 52.0   
10  CAPTURA      ABADEJO                           37                 37.0   
11  CAPTURA      ABADEJO                         2280               2280.0   
12  CAPTURA      ABADEJO                           13                 13.0   
13  CAPTURA      ABADEJO                           77                 77.0   
14  CAPTURA      ABADEJO                        40016              40016.0   
15  CAPTURA      ABADEJO                          221                221.0   
16  CAPTURA      ABADEJO                         1100               1210.0   
17  CAPTURA      ABADEJO                          155                155.0   
18  CAPTURA      ABADEJO                          240                240.0   
19  CAPTURA      ABADEJO                          430                430.0   

    VALOR PESOS NOMBRE PRINCIPAL  CLAVE DE ENTIDAD  CLAVE DE OFICINA  
0     3119.6928             MERO               2.0             207.0  
1         46424             MERO               2.0             207.0  
2     159567.13             MERO               3.0             313.0  
3        616.93             MERO               3.0             312.0  
4     166462.23             MERO               6.0             602.0  
5   159899.1835             MERO               4.0             410.0  
6      611.6664             MERO               4.0             401.0  
7     3525.9621             MERO               4.0             401.0  
8     6578.2216         

## DATA CLEANING

### DataFrame Insight

In [70]:
df.shape

(312752, 12)

In [71]:
df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 312752 entries, 0 to 63235
Data columns (total 12 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   EJERCICIO FISCAL              251440 non-null  float64
 1   ENTIDAD FEDERATIVA            251440 non-null  object 
 2   NOMBRE OFICINA                251440 non-null  object 
 3   MES DE CORTE                  251440 non-null  object 
 4   ORIGEN                        251440 non-null  object 
 5   NOMBRE COMUN                  251440 non-null  object 
 6   PESO DESEMBARCADO KILOGRAMOS  251440 non-null  object 
 7   PESO VIVO KILOGRAMOS          251440 non-null  object 
 8   VALOR PESOS                   251440 non-null  object 
 9   NOMBRE PRINCIPAL              251440 non-null  object 
 10  CLAVE DE ENTIDAD              251440 non-null  float64
 11  CLAVE DE OFICINA              251440 non-null  float64
dtypes: float64(3), object(9)
memory usage: 31.0+ 

In [72]:
df.describe(include='all')

EJERCICIO FISCAL   ENTIDAD FEDERATIVA NOMBRE OFICINA MES DE CORTE  \
count      251440.000000               251440         251440       251440   
unique               NaN                   31            157           12   
top                  NaN  BAJA CALIFORNIA SUR       ENSENADA        MARZO   
freq                 NaN                31614           5571        22339   
mean         2018.489449                  NaN            NaN          NaN   
std             1.124732                  NaN            NaN          NaN   
min          2017.000000                  NaN            NaN          NaN   
25%          2017.000000                  NaN            NaN          NaN   
50%          2018.000000                  NaN            NaN          NaN   
75%          2020.000000                  NaN            NaN          NaN   
max          2020.000000                  NaN            NaN          NaN   

         ORIGEN NOMBRE COMUN PESO DESEMBARCADO KILOGRAMOS  \
count    251440       251440                       251440   
unique        2          675                        67919   
top     CAPTURA   TILAPIA SE                          100   
freq     230078        10935                         1438   
mean        NaN          NaN                          NaN   
std         NaN          NaN                          NaN   
min         NaN          NaN                          NaN   
25%         NaN          NaN                          NaN   
50%         NaN          NaN                          NaN   
75%         NaN          NaN                          NaN   
max         NaN          NaN                          NaN   

       PESO VIVO KILOGRAMOS VALOR PESOS NOMBRE PRINCIPAL  CLAVE DE ENTIDAD  \
count                251440      251440           251440     251440.000000   
unique                77851      170970               57               NaN   
top                       0       6,000            OTRAS               NaN   
freq                   1250         247            48999               NaN   
mean                    NaN         NaN              NaN         17.789373   
std                     NaN         NaN              NaN         10.575871   
min                     NaN         NaN              NaN          1.000000   
25%                     NaN         NaN              NaN          4.000000   
50%                     NaN         NaN              NaN         20.000000   
75%                     NaN         NaN              NaN         27.000000   
max                     NaN         NaN              NaN         32.000000   

        CLAVE DE OFICINA  
count      251440.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean         1784.014067  
std          1058.024604  
min           101.000000  
25%           410.000000  
50%          2004.000000  
75%          2705.000000  
max          3201.000000

- note how we have 312752 rows but only 251440 from those are not null.
- Summary stadistics were calculated for 'EJERCICIO FISCAL','CLAVE DE OFICINA' and 'CLAVE DE ENTIDAD' althouh they should be categorical values. 
- The stadistical descriptors were not calculated for the numerical columns, meaning there most be mixed data oro strings; this will be solved in the following cells. 

### renaming columns and dropping columns

In [73]:
#rename columns and data frame and deletion of the 2 last columns which will not be used. 
df = df.rename(columns={'ENTIDAD FEDERATIVA':'ESTADO','EJERCICIO FISCAL':'FECHA','NOMBRE OFICINA':'OFICINA',"NOMBRE PRINCIPAL":'FAMILIA',"NOMBRE COMUN":'ESPECIE'}).drop(columns=['CLAVE DE ENTIDAD','CLAVE DE OFICINA'])
df.head(10)

FECHA               ESTADO         OFICINA MES DE CORTE   ORIGEN  ESPECIE  \
0  2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO  CAPTURA  ABADEJO   
1  2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO  CAPTURA  ABADEJO   
2  2017.0  BAJA CALIFORNIA SUR  BAHIA ASUNCION       AGOSTO  CAPTURA  ABADEJO   
3  2017.0  BAJA CALIFORNIA SUR   SANTA ROSALIA    NOVIEMBRE  CAPTURA  ABADEJO   
4  2017.0               COLIMA         TECOMAN      FEBRERO  CAPTURA  ABADEJO   
5  2017.0             CAMPECHE          ATASTA    NOVIEMBRE  CAPTURA  ABADEJO   
6  2017.0             CAMPECHE        CAMPECHE        ENERO  CAPTURA  ABADEJO   
7  2017.0             CAMPECHE        CAMPECHE        ABRIL  CAPTURA  ABADEJO   
8  2017.0             CAMPECHE        CAMPECHE         MAYO  CAPTURA  ABADEJO   
9  2017.0             CAMPECHE       CHAMPOTON         MAYO  CAPTURA  ABADEJO   

  PESO DESEMBARCADO KILOGRAMOS PESO VIVO KILOGRAMOS  VALOR PESOS FAMILIA  
0                           94                 94.0    3119.6928    MERO  
1                         1400               1540.0        46424    MERO  
2                         4397               4397.0    159567.13    MERO  
3                           17                 17.0       616.93    MERO  
4                         4587               4587.0    166462.23    MERO  
5                         4406               4406.0  159899.1835    MERO  
6                           18                 18.0     611.6664    MERO  
7                          103                103.0    3525.9621    MERO  
8                          189                189.0    6578.2216    MERO  
9                           52                 52.0     1767.647    MERO

### Treating with missing values and Mixed Dtypes

In [74]:
print("NaN FECHA",df['FECHA'].isnull().sum())
print("NaN PESO DESEMBARCADO KILOGRAMOS",df['PESO DESEMBARCADO KILOGRAMOS'].isnull().sum())
print("NaN PESO VIVO KILOGRAMOS",df['PESO VIVO KILOGRAMOS'].isnull().sum())
print("NaN VALOR PESOS",df['VALOR PESOS'].isnull().sum())

NaN FECHA 61312
NaN PESO DESEMBARCADO KILOGRAMOS 61312
NaN PESO VIVO KILOGRAMOS 61312
NaN VALOR PESOS 61312


In [75]:
#drop rows in which all data is missing and reset the index
df=df.dropna(how='all').reset_index(drop=True)
display(df)
# transforming string or float values to numbers
cols = ['PESO VIVO KILOGRAMOS', 'PESO DESEMBARCADO KILOGRAMOS', 'VALOR PESOS','FECHA']
df[cols] = df[cols].apply(pd.to_numeric, errors='coerce', axis=1)
#replace NaN with O values
df['FECHA'] = df['FECHA'].replace(np.nan, 0)
df['PESO DESEMBARCADO KILOGRAMOS'] = df['PESO DESEMBARCADO KILOGRAMOS'].replace(np.nan, 0)
df['PESO VIVO KILOGRAMOS']= df['PESO VIVO KILOGRAMOS'].replace(np.nan, 0)
df['VALOR PESOS'] = df['VALOR PESOS'].replace(np.nan, 0)

FECHA               ESTADO         OFICINA MES DE CORTE       ORIGEN  \
0       2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO      CAPTURA   
1       2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO      CAPTURA   
2       2017.0  BAJA CALIFORNIA SUR  BAHIA ASUNCION       AGOSTO      CAPTURA   
3       2017.0  BAJA CALIFORNIA SUR   SANTA ROSALIA    NOVIEMBRE      CAPTURA   
4       2017.0               COLIMA         TECOMAN      FEBRERO      CAPTURA   
...        ...                  ...             ...          ...          ...   
251435  2020.0               PUEBLA          PUEBLA       AGOSTO  ACUACULTURA   
251436  2020.0               PUEBLA          PUEBLA      OCTUBRE  ACUACULTURA   
251437  2020.0               PUEBLA          PUEBLA    NOVIEMBRE  ACUACULTURA   
251438  2020.0               PUEBLA          PUEBLA    DICIEMBRE  ACUACULTURA   
251439  2020.0             VERACRUZ        VERACRUZ        MARZO  ACUACULTURA   

          ESPECIE PESO DESEMBARCADO KILOGRAMOS PESO VIVO KILOGRAMOS  \
0         ABADEJO                           94                 94.0   
1         ABADEJO                         1400               1540.0   
2         ABADEJO                         4397               4397.0   
3         ABADEJO                           17                 17.0   
4         ABADEJO                         4587               4587.0   
...           ...                          ...                  ...   
251435  TRUCHA SE                        1,984                1,984   
251436  TRUCHA SE                        3,100                3,100   
251437  TRUCHA SE                          484                  484   
251438  TRUCHA SE                       16,615               16,615   
251439  TRUCHA SE                          248                  248   

       VALOR PESOS FAMILIA  
0        3119.6928    MERO  
1            46424    MERO  
2        159567.13    MERO  
3           616.93    MERO  
4        166462.23    MERO  
...            ...     ...  
251435     150,139  TRUCHA  
251436     211,463  TRUCHA  
251437      36,477  TRUCHA  
251438   1,208,277  TRUCHA  
251439      17,340  TRUCHA  

[251440 rows x 10 columns]

In [76]:
print("NaN FECHA",df['FECHA'].isnull().sum())
print("NaN PESO DESEMBARCADO KILOGRAMOS",df['PESO DESEMBARCADO KILOGRAMOS'].isnull().sum())
print("NaN PESO VIVO KILOGRAMOS",df['PESO VIVO KILOGRAMOS'].isnull().sum())
print("NaN VALOR PESOS",df['VALOR PESOS'].isnull().sum())


NaN FECHA 0
NaN PESO DESEMBARCADO KILOGRAMOS 0
NaN PESO VIVO KILOGRAMOS 0
NaN VALOR PESOS 0


In [77]:
print("NaN FECHA:",df['FECHA'].isnull().values.any())
print("NaN PESO DESEMBARCADO KILOGRAMOS:",df['PESO DESEMBARCADO KILOGRAMOS'].isnull().values.any())
print("NaN PESO VIVO KILOGRAMOS:",df['PESO VIVO KILOGRAMOS'].isnull().values.any())
print("NaN VALOR PESOS:",df['VALOR PESOS'].isnull().values.any())

NaN FECHA: False
NaN PESO DESEMBARCADO KILOGRAMOS: False
NaN PESO VIVO KILOGRAMOS: False
NaN VALOR PESOS: False


In [78]:
#transform numerical colums to int64
columns=['PESO VIVO KILOGRAMOS', 'PESO DESEMBARCADO KILOGRAMOS', 'VALOR PESOS']
df[columns]=df[columns].astype(np.int64)
#transform objects to categories
for col in ['FECHA','ESTADO', 'MES DE CORTE', 'ORIGEN','ESPECIE','FAMILIA']:
    df[col] = df[col].astype('category')

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251440 entries, 0 to 251439
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype   
---  ------                        --------------   -----   
 0   FECHA                         251440 non-null  category
 1   ESTADO                        251440 non-null  category
 2   OFICINA                       251440 non-null  object  
 3   MES DE CORTE                  251440 non-null  category
 4   ORIGEN                        251440 non-null  category
 5   ESPECIE                       251440 non-null  category
 6   PESO DESEMBARCADO KILOGRAMOS  251440 non-null  int64   
 7   PESO VIVO KILOGRAMOS          251440 non-null  int64   
 8   VALOR PESOS                   251440 non-null  int64   
 9   FAMILIA                       251440 non-null  category
dtypes: category(6), int64(3), object(1)
memory usage: 9.4+ MB


In [80]:
df_fishing_2017_2020=df
df_fishing_2017_2020

FECHA               ESTADO         OFICINA MES DE CORTE       ORIGEN  \
0       2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO      CAPTURA   
1       2017.0      BAJA CALIFORNIA      SAN FELIPE        JULIO      CAPTURA   
2       2017.0  BAJA CALIFORNIA SUR  BAHIA ASUNCION       AGOSTO      CAPTURA   
3       2017.0  BAJA CALIFORNIA SUR   SANTA ROSALIA    NOVIEMBRE      CAPTURA   
4       2017.0               COLIMA         TECOMAN      FEBRERO      CAPTURA   
...        ...                  ...             ...          ...          ...   
251435  2020.0               PUEBLA          PUEBLA       AGOSTO  ACUACULTURA   
251436  2020.0               PUEBLA          PUEBLA      OCTUBRE  ACUACULTURA   
251437  2020.0               PUEBLA          PUEBLA    NOVIEMBRE  ACUACULTURA   
251438  2020.0               PUEBLA          PUEBLA    DICIEMBRE  ACUACULTURA   
251439  2020.0             VERACRUZ        VERACRUZ        MARZO  ACUACULTURA   

          ESPECIE  PESO DESEMBARCADO KILOGRAMOS  PESO VIVO KILOGRAMOS  \
0         ABADEJO                            94                    94   
1         ABADEJO                          1400                  1540   
2         ABADEJO                          4397                  4397   
3         ABADEJO                            17                    17   
4         ABADEJO                          4587                  4587   
...           ...                           ...                   ...   
251435  TRUCHA SE                             0                     0   
251436  TRUCHA SE                             0                     0   
251437  TRUCHA SE                           484                   484   
251438  TRUCHA SE                             0                     0   
251439  TRUCHA SE                           248                   248   

        VALOR PESOS FAMILIA  
0              3119    MERO  
1             46424    MERO  
2            159567    MERO  
3               616    MERO  
4            166462    MERO  
...             ...     ...  
251435            0  TRUCHA  
251436            0  TRUCHA  
251437            0  TRUCHA  
251438            0  TRUCHA  
251439            0  TRUCHA  

[251440 rows x 10 columns]

In [81]:
df_fishing_2017_2020.describe(include="all")

FECHA               ESTADO   OFICINA MES DE CORTE   ORIGEN  \
count   251440.0               251440    251440       251440   251440   
unique       4.0                   31       157           12        2   
top       2017.0  BAJA CALIFORNIA SUR  ENSENADA        MARZO  CAPTURA   
freq     64386.0                31614      5571        22339   230078   
mean         NaN                  NaN       NaN          NaN      NaN   
std          NaN                  NaN       NaN          NaN      NaN   
min          NaN                  NaN       NaN          NaN      NaN   
25%          NaN                  NaN       NaN          NaN      NaN   
50%          NaN                  NaN       NaN          NaN      NaN   
75%          NaN                  NaN       NaN          NaN      NaN   
max          NaN                  NaN       NaN          NaN      NaN   

           ESPECIE  PESO DESEMBARCADO KILOGRAMOS  PESO VIVO KILOGRAMOS  \
count       251440                  2.514400e+05          2.514400e+05   
unique         675                           NaN                   NaN   
top     TILAPIA SE                           NaN                   NaN   
freq         10935                           NaN                   NaN   
mean           NaN                  8.735755e+03          1.723146e+04   
std            NaN                  2.201736e+05          3.462989e+05   
min            NaN                  0.000000e+00          0.000000e+00   
25%            NaN                  0.000000e+00          0.000000e+00   
50%            NaN                  1.510000e+02          2.640000e+02   
75%            NaN                  8.510000e+02          1.725000e+03   
max            NaN                  5.789263e+07          5.789263e+07   

         VALOR PESOS FAMILIA  
count   2.514400e+05  251440  
unique           NaN      57  
top              NaN   OTRAS  
freq             NaN   48999  
mean    3.034289e+05     NaN  
std     6.115676e+06     NaN  
min     0.000000e+00     NaN  
25%     0.000000e+00     NaN  
50%     0.000000e+00     NaN  
75%     2.298000e+04     NaN  
max     1.161669e+09     NaN

### Data Saving

In [84]:
df_fishing_2017_2020.to_csv("data/clean_df.csv", index=False)